crates/sui-network/src/state_sync/mod.rs

```diff
@@ -411,13 +418,29 @@ where
                 assert_eq!(checkpoint.previous_digest(), Some(previous_digest));
             }
 
-            // Otherwise stick it with the other unprocessed checkpoints and we can try to sync the missing
-            // ones
-            self.peer_heights
-                .write()
-                .unwrap()
-                .insert_checkpoint(checkpoint.into_inner());
-            warn!("Consensus gave us too new of a checkpoint");
+            // See if the missing checkpoints are already in our store and quickly update our
+            // watermarks
+            for sequence_number in next_sequence_number..=checkpoint.sequence_number() {
+                if let Some(checkpoint) = self
+                    .store
+                    .get_checkpoint_by_sequence_number(sequence_number)
+                    .expect("store operation should not fail")
+                {
+                    self.store
+                        .insert_checkpoint(checkpoint.clone())
+                        .expect("store operation should not fail");
+                    self.store
+                        .update_highest_synced_checkpoint(&checkpoint)
+                        .expect("store operation should not fail");
+                    self.metrics
+                        .set_highest_verified_checkpoint(checkpoint.sequence_number());
+                    self.metrics
+                        .set_highest_synced_checkpoint(checkpoint.sequence_number());
```

Comment:
how do we detect/monitor holes if we bump watermarks for any seq num?

Reply:
I've changed this so that if we do run into a hole we'll stop bumping the watermark.